streams/src/main/java/org/apache/kafka/streams/kstream/Named.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.common.errors.InvalidTopicException;
+import org.apache.kafka.common.internals.Topic;
+import org.apache.kafka.streams.errors.InvalidProcessorException;
+
+import java.util.Objects;
+
+public class Named implements NamedOperation<Named> {
+
+    protected String name;
+
+    protected Named(final Named named) {
+        this(Objects.requireNonNull(named, "named can't be null").name);
+    }
+
+    protected Named(final String name) {
+        this.name = name;
+        if (name != null) {
+            // Validate the specified processor name.
+            try {
+                Topic.validate(name);
```

Comment:
@fhussonnois thinking about this some more, what is the motivation for doing a validation here for processor names?  

When Streams starts up the `AdminClient` will attempt to create any internal topics which will get validated at that point, so we don't need this check up front.

\cc @guozhangwang

Reply:
@bbejeck, actually I've followed the same logic than for the class Materialized. In general, we should try to throw the exception as soon as possible. Also, here the exception is thrown during the build of the topology (a bad name will be detected during unit tests). If we do not check the name in the Named/Materialized classes then an nivalid name will be only detected during runtime. Depending of the semantic we would like to have, we can consider than an invalid name is topology exception or an invalid topic exception ?

Also, I'm also in favor to duplicate `Topic.validate` as this will allow to not tie the operation name with the topic name in the exception message.

Reply:
You have a point about checking via unit tests, but when the `AdminClient` attempts to create the topic, the check is against the full name.

The check here only looks at the user-supplied name, which is only part of the topic name so that we could have a situation, although admittedly rare, where the unit tests pass, the full name fails.

So I'm inclined not to have the check here or in the `Materialized` class.  Let's see what others think.
\cc @guozhangwang @mjsax @vvcephei @ableegoldman

Reply:
I agree, that we should check when using AdminClient -- however, I also agree that we should check early if possible. (I don't see an issue with double checking.) Sometimes a name is used as part of a topic name, and we can check for invalid characters here already. Could we add a flag like "usedInTopicName" and do different check accordingly? On the other hand, I don't see a big issue with enforcing the topic naming conventions to processor names, too, even if it's just a processor name and not used as part of a topic name.

Reply:
I've thought about this some more and comments from @mjsax and @fhussonnois have convinced me the checking earlier is a good idea as well.

But I still think we should duplicate the current logic and put the name checking logic in the `Named` class.

Reply:
To re-start this thread, I also feel like we should have our own check for Named operations to use.
* we may want to make operation names more restrictive than topic names, for example to prevent collisions with automatically-generated partition names
* the topic validation throws an exception that mentions the name is "an invalid topic name". This statement is nonsense if I'm naming an operator. We should throw an exception that says it's an "invalid operator name" or similar.